### PR TITLE
feat(train): add fsdp training option

### PR DESCRIPTION
**Pull request by zaglc**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
This PR adds interface for FSDP, which can be a subsitution of ZeRO1 and is compatible with TENSOR parallel. Running FSDP with PIPELINE parallel is not supported in this PR.

## Modification
+ add another group initializer `Initializer_Zero3_dp` in case that ZeRO3 parallel size is less than DP size and all-reduce is needed;
+ add an simplified optimizer `FSDPadaptOptimizer` because FSDP module will automatically accompolish most of the communication;
+ add interface for FSDP model storing in order to be compatible with current framework;
+ add interface of FSDP and warp the original model;

## Use cases (Optional)
If you want to train the model with FSDP, set the use_fsdp flag in parallel dict in 7B_sft.py like this:
```
parallel = dict(
    zero1=-1,
    pipeline=dict(size=1, interleaved_overlap=True),
    tensor=2,
    sequence_parallel=False,
    use_fsdp = True,
)
```

## Checklist
+ The PR requires more test for some reasons such as loss value alignment with original parallelized training method
+ May commit other changes basing on this PR in the future

### CODE REVIEW

**Comment by zaglc** [comment]
In last commit, the mix precision training was incomplete, training loss on a small dataset vary like this (zero1 represents training without fsdp):
图片名称
after commit 31d2a29, training loss align well:
图片名称

File: configs/7B_sft.py [hunk 1]
```diff
@@ -1,42 +1,48 @@
 JOB_NAME = "7b_train"
 DO_ALERT = False
 
-SEQ_LEN = 2048
-HIDDEN_SIZE = 4096
+SEQ_LEN = 256
```
Comment (sunpengsdu) [inline configs/7B_sft.py:4]: 最好不要更改这个py，新建一个fsdp测试用的config为好

Comment (zaglc) [reply configs/7B_sft.py:4]: 好的

File: internlm/train/training_internlm.py [hunk 2]
```diff
@@ -92,6 +103,24 @@ def initialize_model():
     return model
 
 
+def warp_FSDP_model(model: Union[nn.Module, nn.ModuleList]):
```
Comment (huangting4201) [inline internlm/train/training_internlm.py:106]: 这里是不是warp拼错了

File: internlm/utils/model_checkpoint.py [hunk 3]
```diff
@@ -154,6 +154,26 @@ def get_model_topology(model):
...
+def get_state_dict(model):
+    """
+    Only used for FSDP module saving.
+    It's a warper of model.state_dict() and with the context of 'FSDP.state_dict_type', the sharded parameter
+    (saved as model.flat_param_xx in sharded FSDP module) will be gathered at every gpu.
+    'offload_to_cpu' means that the model states are to be offloaded to cpu chunk by chunk, avoiding OOM in gpu
+
+    """
+    from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+    from torch.distributed.fsdp import FullStateDictConfig, StateDictType
```
Comment (huangting4201) [inline internlm/utils/model_checkpoint.py:166]: import放在文件顶部吧

File: internlm/core/context/parallel_context.py [hunk 4]
```diff
@@ -478,6 +478,8 @@ def init_parallel_groups(self):
         initializers.append(pgroup_initializer.Initializer_Model(*initializer_args))
         initializers.append(pgroup_initializer.Initializer_Tensor(*initializer_args))
         initializers.append(pgroup_initializer.Initializer_Zero1(*initializer_args))
+        if self.config.parallel.use_fsdp:
```
Comment (huangting4201) [inline internlm/core/context/parallel_context.py:481]: 这里得改成`if self.config.parallel.get("use_fsdp", False)`，不然ci train会key error

File: internlm/solver/optimizer/hybrid_zero_optim.py [hunk 5]
```diff
@@ -81,6 +81,211 @@ def clip_grad_norm(self):
...
+                break
+            norm_groups[group_name] = norm_group
+
+        loss_scale = float(self.loss_scale.item())  # backup
+        self.grad_scaler.update(found_inf)
+        if found_inf:
+            if gpc.is_rank_for_log():
+                logger.warning("Overflow occurs, please check it.")
+            self.zero_grad()
+            return False, None
```
Comment (huangting4201) [inline internlm/solver/optimizer/hybrid_zero_optim.py:190]: 181行要去掉break，因为found_inf的情况下，norm_groups也是要返回的，所以这里190行也要改成`return False, norm_groups`

File: internlm/solver/optimizer/hybrid_zero_optim.py [hunk 6]
```diff
@@ -81,6 +81,211 @@ def clip_grad_norm(self):
...
+            grad_fp32 = [p.grad.to(dtype) for p in fp16_params]
+
+            device = self._fp32_param_tensor_groups[group_idx][0].device
+            for p, g in zip(self._fp32_param_tensor_groups[group_idx], grad_fp32):
+                p.grad = g.to(device)
+
+        # unscale
+        self._unscale_and_clip_grads(list(global_norm_groups.values()), loss_scale)
+
+        self.optim.step()
```
Comment (huangting4201) [inline internlm/solver/optimizer/hybrid_zero_optim.py:211]: step之后好像少了release the fp32 grad

Comment (zaglc) [reply internlm/solver/optimizer/hybrid_zero_optim.py:211]: 好的，我也再检查一下，下次commit一并改

File: internlm/train/training_internlm.py (hunk 2 above)
Comment (zaglc) [reply internlm/train/training_internlm.py:106]: 哦哦是的，本来想写wrap的

File: internlm/utils/model_checkpoint.py (hunk 3 above)
Comment (zaglc) [reply internlm/utils/model_checkpoint.py:166]: 好的，我把它放上去

**Review by huangting4201 — APPROVED** [review]

**Comment by huangting4201** [comment]
这个pr的title可能要再规范具体一点

File: train.py [hunk 7]
```diff
@@ -110,6 +111,9 @@ def main(args):
     # initialize and resume train state
     train_state = TrainState(gpc.config, train_dl.batch_sampler)
 
+    # if fsdp enabled, warp the model
+    model = wrap_FSDP_model(model)
```
Comment (huangting4201) [inline train.py:115]: mark一下，私仓也要相应更新

Comment (sunpengsdu) [reply train.py:115]: @huangting4201 这个能放在init model那个函数吗

File: internlm/utils/model_checkpoint.py [hunk 8]
```diff
@@ -154,11 +156,44 @@ def get_model_topology(model):
...
+
+    return (missing_k, unexpected_keys)
+
+
 def try_load_internlm_ckpt(ckpt_mm, load_info, train_state: TrainState):
     load_content_str = ""
     load_ckpt_folder = load_info["path"]
     load_content: CheckpointLoadMask = load_info["content"]
-
+    print(load_ckpt_folder, load_content)
```
Comment (sunpengsdu) [inline internlm/utils/model_checkpoint.py:196]: 避免用print? 用log?

File: internlm/core/context/process_group_initializer.py [hunk 9]
```diff
@@ -416,3 +421,62 @@ def init_dist_group(self, use_cpu: bool = False):
...
+        world_size (int): Size of whole communication world.
+        data_parallel_size (int): Size of data parallel.
+        pipeline_parallel_size (int): Size of pipeline parallel.
+        tensor_parallel_size (int): Size of tensor parallel.
+        zero1_parallel_size (int): Size of zero1 parallel.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        assert self.data_parallel_size % self.zero1_parallel_size == 0
```
Comment (sunpengsdu) [inline internlm/core/context/process_group_initializer.py:440]: fsdp和zero1这个数值会不会有冲突

Comment (sunpengsdu) [reply internlm/core/context/process_group_initializer.py:440]: 是不是只能支持-1为好？

File: internlm/solver/optimizer/hybrid_zero_optim.py [hunk 10]
```diff
@@ -81,6 +81,208 @@ def clip_grad_norm(self):
         pass
 
 
+class FSDPadaptOptimizer(BaseOptimizer):
```
Comment (sunpengsdu) [inline internlm/solver/optimizer/hybrid_zero_optim.py:84]: fsdp的optimizer要不要单独一个.py文件比较好？和线上用的在物理上切割开

File: internlm/core/context/process_group_initializer.py (hunk 9 above)
Comment (sunpengsdu) [reply internlm/core/context/process_group_initializer.py:440]: 似乎好像是的

File: internlm/utils/model_checkpoint.py (hunk 8 above)
Comment (zaglc) [reply internlm/utils/model_checkpoint.py:196]: 抱歉这个是我忘记删除了，之后会改掉

File: train.py (hunk 7 above)
Comment (zaglc) [reply train.py:115]: @sunpengsdu 可以的，原来写在外面是因为ckpt的问题，后来ckpt load的方式调整了，就可以写在里面了，下次提交一并改

File: internlm/core/context/process_group_initializer.py (hunk 9 above)
Comment (huangting4201) [reply internlm/core/context/process_group_initializer.py:440]: 为啥只能支持zero1 -1呢

Comment (zaglc) [reply internlm/core/context/process_group_initializer.py:440]: zero3_dp那个通信组就是为了支持zerosize<dpsize的情况的，目前zero1=-1精度是没问题的，但zero1 < dpsize的时候有点对不上，正在排查

File: internlm/solver/optimizer/hybrid_zero_optim.py (hunk 10 above)
Comment (huangting4201) [reply internlm/solver/optimizer/hybrid_zero_optim.py]: 已更新https://github.com/InternLM/InternLM/pull/293/commits/f3f2511e74d7fe0271fbf450ae85579acdde05b8

File: train.py (hunk 7 above)
Comment (huangting4201) [reply train.py]: 已更新https://github.com/InternLM/InternLM/pull/293/commits/f3f2511e74d7fe0271fbf450ae85579acdde05b8

File: internlm/core/context/process_group_initializer.py (hunk 9 above)
Comment (huangting4201) [reply internlm/core/context/process_group_initializer.py:440]: 已更新[056996f](https://github.com/InternLM/InternLM/pull/293/commits/056996f8b31e8c6b42ec31784e9f72dd26c22a11)

File: internlm/utils/model_checkpoint.py [hunk 11]
```diff
@@ -353,15 +352,18 @@ def load_model_checkpoint(folder, model):
...
     if gpc.config.parallel.use_fsdp:
         assert (
-            zo_size == max_zo + 1
-        ), f"The weights are save for {max_zo+1} FSDP shards , while current has {zo_size} FSDP shards"
+            dp_size == max_zo + 1
+        ), f"The weights are save for {max_zo+1} FSDP shards , while current has {dp_size} FSDP shards"
 
     if gpc.config.parallel.use_fsdp:
-        should_load_name = f"model_tp{tp_rank}_pp{pp_rank}_zo{zo_rank}.pt"
+        should_load_name = f"model_tp{tp_rank}_pp{pp_rank}_zo{dp_rank}.pt"
```
Comment (huangting4201) [inline internlm/utils/model_checkpoint.py:359]: 文件名要不要改成`model_tp{tp_rank}_pp{pp_rank}_dp{dp_rank}`啊

**Comment by sunpengsdu** [comment]
现在的主要问题就还是不开activation ckpt就会err吧?

**Comment by huangting4201** [comment]
是的

> 现在的主要问题就还是不开activation ckpt就会err吧?

是的

File: train.py [hunk 12]
```diff
@@ -289,6 +289,8 @@ def main(args):
 
 
 if __name__ == "__main__":
+    assert torch.__version__ >= "2.0.1", f"requires torch>=2.0.1 but current version is {torch.__version__}"
```
Comment (huangting4201) [inline train.py:292]: 这个校验应该是在开启fsdp时才判断

Comment (zaglc) [reply train.py:304]: 好的

File: internlm/train/training_internlm.py [hunk 13]
```diff
@@ -117,18 +109,23 @@ def initialize_model():
...
         transformer_wrap_policy = functools.partial(
-            transformer_auto_wrap_policy, transformer_layer_cls={
-                PackedFlashBaseLayer1D, 
-                PackedFlashInternLm1D, 
-                MHA, 
-                FlashCrossAttention, 
-                FlashSelfAttention, 
-                RMSNorm}
+            transformer_auto_wrap_policy,
+            transformer_layer_cls={PackedFlashBaseLayer1D, PackedFlashInternLm1D, MHA, RMSNorm},
```
[the remark this answers is not in the document]
Comment (sunpengsdu) [reply internlm/train/training_internlm.py:126]: 是不是只保留更细的粒度就行？MHA，ScaleColumnParallelLinear，RowParallelLinearTorch，ColumnParallelLinearTorch等等，底下有parameter本质都是 nn.Linear

**Comment by sunpengsdu** [comment]
FSDP和混合精度会有冲突吗？

**Comment by zaglc** [comment]
> FSDP和混合精度会有冲突吗？

鹏哥说的混合精度是optimizer里的混合精度吗？fsdp optimizer里面仿照hybird zero手动实现了混合精度

**Comment by huangting4201** [comment]
> FSDP和混合精度会有冲突吗？

混合精度这个我测试了一下把norm改成fp32，是能正常跑的

File: internlm/train/training_internlm.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (huangting4201) [reply internlm/train/training_internlm.py]: 已更新
